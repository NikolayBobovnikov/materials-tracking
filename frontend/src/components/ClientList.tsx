import React from 'react';
import { graphql, useLazyLoadQuery } from 'react-relay';
import { 
  Paper, 
  Table, 
  TableBody, 
  TableCell, 
  TableContainer, 
  TableHead, 
  TableRow, 
  Typography,
  Box,
  Button
} from '@mui/material';
import type { ClientListQuery } from './../__generated__/ClientListQuery.graphql';
import { ResponseType } from '../utils/types';
import { validateRelayData, validateRelayConnection } from '../utils/dataValidator';

// Define our query
const query = graphql`
  query ClientListQuery($first: Int, $after: String) {
    clients(first: $first, after: $after) {
      edges {
        node {
          id
          name
          markup_rate
        }
        cursor
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;

const ClientList: React.FC = () => {
  const [first] = React.useState(10);
  const [after, setAfter] = React.useState<string | null>(null);
  
  const data = useLazyLoadQuery<ResponseType<ClientListQuery>>(
    query, 
    { first, after }
  );

  // Validate data structure at runtime
  validateRelayData(data, 'ClientList');
  const clientsValid = validateRelayConnection(data.clients, 'clients', 'ClientList');

  const loadMore = (): void => {
    if (data.clients?.pageInfo.hasNextPage) {
      setAfter(data.clients.pageInfo.endCursor);
    }
  };

  if (!clientsValid || !data.clients || !data.clients.edges || data.clients.edges.length === 0) {
    return (
      <Box sx={{ mt: 2 }}>
        <Typography variant="h6" gutterBottom>Clients</Typography>
        <Typography>No clients found.</Typography>
      </Box>
    );
  }

  return (
    <Box sx={{ mt: 2 }}>
      <Typography variant="h6" gutterBottom>Clients</Typography>
      <TableContainer component={Paper}>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell>ID</TableCell>
              <TableCell>Name</TableCell>
              <TableCell>Markup Rate</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {data.clients.edges.map((edge) => {
              if (!edge || !edge.node) return null;
              const node = edge.node;
              return (
                <TableRow key={node.id}>
                  <TableCell>{node.id}</TableCell>
                  <TableCell>{node.name}</TableCell>
                  <TableCell>{node.markup_rate}</TableCell>
                </TableRow>
              );
            })}
          </TableBody>
        </Table>
      </TableContainer>
      
      {data.clients.pageInfo.hasNextPage && (
        <Button 
          variant="outlined" 
          onClick={loadMore} 
          sx={{ mt: 2 }}
        >
          Load More
        </Button>
      )}
    </Box>
  );
};

export default ClientList; 