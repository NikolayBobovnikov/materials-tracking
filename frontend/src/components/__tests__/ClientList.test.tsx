import React from 'react';
import { render } from '@testing-library/react';
import ClientList from '../ClientList';
import { useLazyLoadQuery } from 'react-relay';

// Mock the generated type structure
jest.mock('../../__generated__/ClientListQuery.graphql', () => ({
  ClientListQuery: {
    response: {
      clients: {
        edges: [
          {
            node: { id: '1', name: 'Test Client', markup_rate: 1.5 },
            cursor: 'cursor1'
          }
        ],
        pageInfo: { hasNextPage: false, endCursor: null }
      }
    }
  }
}));

// Mock relay hooks
jest.mock('react-relay', () => ({
  useLazyLoadQuery: jest.fn(),
  graphql: jest.fn()
}));

describe('ClientList', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('handles Relay data correctly', () => {
    // Mock the return value of useLazyLoadQuery
    (useLazyLoadQuery as jest.Mock).mockReturnValue({
      clients: {
        edges: [
          { node: { id: '1', name: 'Test Client', markup_rate: 1.5 }, cursor: 'cursor1' }
        ],
        pageInfo: { hasNextPage: false, endCursor: null }
      }
    });
    
    const { container } = render(<ClientList />);
    expect(container).toMatchSnapshot();
  });

  it('calls useLazyLoadQuery with correct parameters', () => {
    // Mock the return value
    (useLazyLoadQuery as jest.Mock).mockReturnValue({
      clients: {
        edges: [],
        pageInfo: { hasNextPage: false, endCursor: null }
      }
    });
    
    render(<ClientList />);
    
    // Verify the hook was called with the expected parameter types
    expect(useLazyLoadQuery).toHaveBeenCalled();
    
    // The second argument should be an object with first and after properties
    const secondArg = (useLazyLoadQuery as jest.Mock).mock.calls[0][1];
    expect(secondArg).toHaveProperty('first');
    expect(typeof secondArg.first).toBe('number');
    
    // Check that after property exists and is either null or a string
    expect(secondArg).toHaveProperty('after');
    // We use a variable and a simple assertion to avoid conditional expect
    const afterTypeIsValid = secondArg.after === null || typeof secondArg.after === 'string';
    expect(afterTypeIsValid).toBe(true);
  });
}); 