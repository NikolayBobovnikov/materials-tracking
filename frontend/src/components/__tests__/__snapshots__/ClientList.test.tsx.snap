// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientList handles Relay data correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-1yuhvjn"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
    >
      Clients
    </h6>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              ID
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Markup Rate
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Test Client
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              1.5
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
